contrib/scoverage/src/ScoverageModule.scala

```diff
@@ -0,0 +1,115 @@
+package mill
+package contrib
+package scoverage
+
+import coursier.{Cache, MavenRepository}
+import mill.api.Result
+import mill.eval.PathRef
+import mill.util.Ctx
+import mill.scalalib.{DepSyntax, JavaModule, Lib, ScalaModule, TestModule, Dep}
+import mill.moduledefs.Cacher
+
+
+/** Adds targets to a [[mill.scalalib.ScalaModule]] to create test coverage reports.
+ *
+ * This module allows you to generate code coverage reports for Scala projects with
+ * [[https://github.com/scoverage Scoverage]] via the
+ * [[https://github.com/scoverage/scalac-scoverage-plugin scoverage compiler plugin]].
+ *
+ * To declare a module for which you want to generate coverage reports you can
+ * Extends the `mill.contrib.scoverage.ScoverageModule` trait when defining your
+ * Module. Additionally, you must define a submodule that extends the
+ * `ScoverageTests` trait that belongs to your instance of `ScoverageModule`.
+ *
+ * {{{
+ * // You have to replace VERSION
+ * import $ivy.`com.lihaoyi::mill-contrib-buildinfo:VERSION`
+ * import mill.contrib.scoverage.ScoverageModule
+ *
+ * Object foo extends ScoverageModule  {
+ *   def scalaVersion = "2.11.8"
+ *   def scoverageVersion = "1.3.1"
+ *
+ *   object test extends ScoverageTests {
+ *     def ivyDeps = Agg(ivy"org.scalatest::scalatest:3.0.5")
+ *     def testFrameworks = Seq("org.scalatest.tools.Framework")
+ *   }
+ * }
+ * }}}
+ *
+ * In addition to the normal tasks available to your Scala module, Scoverage
+ * Modules introduce a few new tasks and changes the behavior of an existing one.
+ *
+ * - mill foo.scoverage.compile      # compiles your module with test instrumentation
+ *                                 # (you don't have to run this manually, running the test task will force its invocation)
+ *
+ * - mill foo.test                   # tests your project and collects metrics on code coverage
+ * - mill foo.scoverage.htmlReport   # uses the metrics collected by a previous test run to generate a coverage report in html format
+ *
+ * The measurement data is available at `out/foo/scoverage/data/`,
+ * And the html report is saved in `out/foo/scoverage/htmlReport/`.
+ */
+trait ScoverageModule extends ScalaModule { outer: ScalaModule =>
```

Comment:
Any reason for extending ScalaModule here ? Considering you have a self-type, extending `mill.define.Module` should be sufficient to define tasks

Reply:
I'd rather keep the `extends` to make it easier to use `ScoverageModule`. I assume the motivation for the self-type it the named handle.

Reply:
I think it might create problems when trying to make it work with scalajs modules (which scoverage supports), as having "extends ScalajsModule with ScoverageModule" could/would have unfortunate results because of inheritance

Reply:
Oh, didn't know that. Though I don't see what the issue could be, as it has to be a `ScalaModule` either way.

Reply:
Maybe, I misunderstood your request. Do you mean to just remove the extends, but keep the `outer: ScalaModule`, or do you suggest to remove both?

Reply:
The former : remove the extend, keep the self-type

Reply:
The extending was to make it easier to use. Since you can write this
```scala
object foo extends ScoverageModule
```
instead of
```scala
object foo extends ScalaModule with ScoverageModule
```
 This shouldn't cause issues with the current implementation:
Because ScalaJsModule already extends ScalaModule
```scala
object foo extends ScalaJsModule with ScoverageModule
```

It will just be a stylistic choice - one that I don't have a strong preference.
FWIW:
`PublishModule extends JavaModule` and its common to see
```scala
object baz extends ScalaModule with PublishModule
```
And ScalaModule already inherits from JavaModule.

Reply:
So it won't cause problems with inheritence, but if others prefer something else stylistically that's fine with me.

Reply:
Right I'm in front of a computer now and had a deeper. Inheritance would be problematic if ScalaJsModule was overriding any of the methods implemented in ScalaModule, which doesn't seem to be the case (I wrongfully thought it was).